File: src/main/java/com/app/Message_Backend/entities/Conversation.java
package com.app.Message_Backend.entities;

import com.app.Message_Backend.listeners.SaveListener;
import org.jetbrains.annotations.NotNull;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.Date;
import java.util.Set;

@EntityListeners( SaveListener.class )
@Entity(name="Conversations")
public class Conversation implements Comparable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;
    @OrderBy("timeSent ASC")
    @OneToMany(mappedBy="conversation", cascade={CascadeType.ALL}, fetch = FetchType.EAGER)
    public Set<Message> messages;
    @ManyToMany(cascade = { CascadeType.MERGE }, fetch = FetchType.EAGER)
    @JoinTable(
            name="Users_Conversations",
            joinColumns = @JoinColumn(name="conversation_id"),
            inverseJoinColumns = @JoinColumn(name="user_id")
    )
    @Column(unique = true)
    public Set<User> users;

    public Conversation() {}

    public Conversation(Long id) {
        this.id = id;
    }

    public Conversation(Set<User> users) {
        this.users = users;
    }

    public Conversation(Set<Message> messages, Set<User> users) {
       this.messages = messages;
       this.users = users;
    }

    public Long getId() {
        return id;
    }

    public Set<Message> getMessages() {
        return messages;
    }

    public void setMessages(Set<Message> messages) {
        this.messages = messages;
    }

    public Set<User> getUsers() {
        return users;
    }

    public void setUsers(Set<User> users) {
        this.users = users;
    }

    @Override
    public int compareTo(@NotNull Object obj) {
        Conversation conv2 = (Conversation) obj;
        int conv1MessageLength = messages.size();
        int conv2MessageLength = conv2.getMessages().size();

        if(conv1MessageLength == 0 && conv2MessageLength == 0) return 0;
        if(conv1MessageLength == 0) return 1;
        if(conv2MessageLength == 0) return -1;

        Date conv1LastMessageSentDate = new ArrayList<Message>(messages).get(conv1MessageLength - 1).getTimeSent();
        Date conv2LastMessageSentDate = new ArrayList<Message>(conv2.getMessages()).get(conv2MessageLength - 1).getTimeSent();
        return conv1LastMessageSentDate.compareTo(conv2LastMessageSentDate) * -1;
    }
}

File: frontend/src/graphql/conversations/Conversation.ts
import Message from "../messages/Message";
import User from "../users/User";


export default class Conversation {
    public readonly id: number;
    public title: string;
    public messages: Array<Message>;
    public users: Array<User>;

    public static fromTitle(title: string): Conversation {
        return new Conversation(0, title, new Array<Message>(), new Array<User>());
    }

    constructor(id: number, title: string, messages: Array<Message>, users: Array<User>) {
        this.id = id;
        this.title = title;
        this.messages = messages;
        this.users = users;
    }

    /**
     * Getter $id
     * @return {number}
     */
	public getId(): number {
		return this.id;
    }
    
    public getTitle(): string {
        return this.title;
    }

    /**
     * Getter $messages
     * @return {Array<Message>}
     */
	public getMessages(): Array<Message> {
		return this.messages;
	}

    /**
     * Getter $users
     * @return {Array<Users>}
     */
	public getUsers() {
		return this.users;
	}
    
    /**
     * Setter $messages
     * @param {Array<Message>} value
     */
	public setMessages(value: Array<Message>) {
		this.messages = value;
	}

    /**
     * Setter $users
     * @param {Array<Users>} value
     */
	public setUsers(value: Array<User>) {
		this.users = value;
	}

}